### ISSUE TITLE
Different transforms applied to CT and label

### ISSUE THREAD

**Issue by simonebonato** [fepegar/torchio#1071]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem summary

I have a CT scan and the corresponding label mask and I want to perform some augmentations on them.

Initially they have a different spacing but after using ToCanonical and Resample(1.0) they become the same; however when I perform some other augmentations and I make a plot where I overlay the label on the image (the same slice ofc), they are not overlapping and the label goes partially on the background.

### Code for reproduction

```python
from utils.plotting import plot_all
import matplotlib.pyplot as plt
import numpy as np
import torchio as tio

subject = tio.Subject(
    CT=tio.ScalarImage(CT_path),
    segmentation=tio.LabelMap(segmentation_path),
)

# Rescale the Intensity of the image
rescaler = tio.transforms.RescaleIntensity((0, 1), percentiles=(0.5, 99.5))
scaled_subject = rescaler(subject)

# Change voxels spacing and bring to canonical Orientation
canonical_spacing_transform = tio.Compose(
    [
        tio.transforms.Resample(1.0),
        tio.transforms.ToCanonical(),
    ]
)
can_spac_subsject = canonical_spacing_transform(scaled_subject)

# Apply data augmentation
augmentation_transform = tio.Compose(
    [
        tio.transforms.RandomAnisotropy(
            p=0.2, include=("CT", "segmentation"), label_keys=("segmentation",)
        ),
        tio.transforms.RandomFlip(
            include=("CT", "segmentation"), label_keys=("segmentation",)
        ),
        tio.transforms.RandomAffine(
            include=("CT", "segmentation"), label_keys=("segmentation",)
        ),
        tio.transforms.RandomNoise(include=("CT"), label_keys=("segmentation",)),
        tio.transforms.RandomBlur(include=("CT"), label_keys=("segmentation",)),
        tio.transforms.RandomBiasField(
            p=0.2, include=("CT"), label_keys=("segmentation",)
        ),
        tio.transforms.RandomElasticDeformation(
            p=0.2, include=("CT", "segmentation"), label_keys=("segmentation",)
        ),
        tio.transforms.RandomGamma(p=0.5, include=("CT"), label_keys=("segmentation",)),
    ]
)

augmented_subject = augmentation_transform(scaled_subject)
```

### Actual outcome

When I plot the label on the CT, this is the result: 
![bad_augmentation](https://user-images.githubusercontent.com/63954877/232458630-ec5ee3fa-8b8a-42f1-ae22-fa260a9f7923.png)

While before the transforms it looks like this: 
![pre_augmentation](https://user-images.githubusercontent.com/63954877/232458585-6b8d865e-fdbf-4808-bfe2-10186f5f6695.png)

The image and the segmentation are the [following](https://kth-my.sharepoint.com/:u:/g/personal/bonato_ug_kth_se/EflzVRAkKeRHmCV5bviFkMQBBNNqCsI6u73yLE552cwb9A?e=clhdnD).

And the slice I printed is CT[500, :, :], same for the label.

### Expected outcome

What I would like to obtain is the label being on top of the CT, and not shifted.

**Comment by romainVala** [comment]
note that in the code you apply other augmentation on scaled_subject but not on can_spac_subsject
not sure if it make any difference

**Comment by simonebonato** [comment]
Thank you for the suggestion.

I tried applying the transforms `RandomFlip` and `RandomAffine`  since they affect the geometry of the CT and the label mask, and not the pixel intensities. 

I found out that the issue comes with `RandomAffine`. I tried changing the hyperparameter "center" (image or origin) and "isotropic" (True or False), in all their combinations, but the misalignment still persists.

I then checked the spacing of the CT and the segmentation mask and I got the following:

<img width="435" alt="image" src="https://user-images.githubusercontent.com/63954877/232503010-1d412a83-0a86-4952-85b3-08c2587474ac.png">

So could it be that the reason could be that they have different spacings? And if I try to give them both the same spacing using `Resample(1.0)` it's also going to change the resolution of the image and at that point I lose the alignment anyway...

Any idea on how to solve this?

**Comment by fepegar** [comment]
Hi, @simonebonato. I would recommend using `CopyAffine` as the firstr transform, as it seems that your segmentations is simply lacking the spatial metadata:

```python
copy_metadata = tio.CopyAffine('CT')
```

However, I just tried that and the transform did nothing in my example! I'll debug this and come back to you.

**Comment by fepegar** [comment]
For visibility, this is somehow not working!

```python
import torch
import torchio as tio
ct = tio.ScalarImage('/tmp/upload_folder/case_01_IMG_CT.nrrd')
seg = tio.LabelMap('/tmp/upload_folder/case_01_segmentation.nrrd')
seg.affine = ct.affine
print(seg.affine)
```

```
array([[-1.,  0.,  0.,  0.],
       [ 0., -1.,  0.,  0.],
       [ 0.,  0.,  1.,  0.],
       [ 0.,  0.,  0.,  1.]])
```

**Comment by fepegar** [comment]
Update: `seg[tio.AFFINE]` gives me the expected output. [Maybe `Image` shouldn't inherit from `dict`](https://treyhunner.com/2019/04/why-you-shouldnt-inherit-from-list-and-dict-in-python/).

**Comment by fepegar** [comment]
Ok, this happens because the affine is not being read correctly as the segmentation hasn't been loaded. That will be addressed. 

https://github.com/fepegar/torchio/blob/0933def2bc5eb61b8c3a15d59944cb09d22eceee/src/torchio/data/image.py#L255-L260

Back to this issue.

**Comment by fepegar** [comment]
```python
import torch
import torchio as tio
ct = tio.ScalarImage('/tmp/upload_folder/case_01_IMG_CT.nrrd')
seg = tio.LabelMap('/tmp/upload_folder/case_01_segmentation.nrrd')

seg.data;  # just to temporarily overcome the bug we mentioned

subject = tio.Subject(ct=ct, seg=seg)

copy_metadata = tio.CopyAffine('ct')
fixed = copy_metadata(subject)
fixed.seg.save('/tmp/upload_folder/case_01_segmentation_fixed.nrrd')
```

The image seems aligned after copying the affine.

<img width="715" alt="image" src="https://user-images.githubusercontent.com/12688084/232570457-8a7812e7-b02c-41a7-8237-331e9088e7b2.png">

**Comment by fepegar** [comment]
> For visibility, this is somehow not working!
> 
> ```python
> import torch
> import torchio as tio
> ct = tio.ScalarImage('/tmp/upload_folder/case_01_IMG_CT.nrrd')
> seg = tio.LabelMap('/tmp/upload_folder/case_01_segmentation.nrrd')
> seg.affine = ct.affine
> print(seg.affine)
> ```
> 
> ```
> array([[-1.,  0.,  0.,  0.],
>        [ 0., -1.,  0.,  0.],
>        [ 0.,  0.,  1.,  0.],
>        [ 0.,  0.,  0.,  1.]])
> ```

This has been fixed in `v0.18.91`.

**Comment by romainVala** [comment]
oh I missed the link of the data
thanks @fepegar for the fix

I think this is a typical errors which is not easy to understand when one is not use to the 3D affine 
@simonebonato  to better understand
try the torchio ploting utility

compare 
subject.plot()
and 
fixed.plot()

taking into account the affine (and thus the voxel size) is important

**Comment by simonebonato** [comment]
Thank you both @fepegar and @romainVala for solving my problem.

And thanks @romainVala I did what you told me and seen the difference.
> compare subject.plot() and fixed.plot()

Next time I will check directly the spacing first before making any operation.
